File: src/rest/internal.ts
import type { OpenAPIHono } from "@hono/zod-openapi";
import { z } from "zod";
import type { AuthEnv } from "../auth/middleware.js";
import { ValidationError } from "../core/errors.js";
import type { EventBus } from "../core/event-bus.js";

const InternalEventBody = z.object({
	type: z.string().regex(/^[a-z]+\.[a-z]+$/),
	data: z.unknown(),
});

/**
 * Internal relay endpoint — accepts change events from the MCP process and
 * replays them onto the HTTP server's event bus for SSE fan-out.
 */
export function registerInternalRoutes(
	api: OpenAPIHono<AuthEnv>,
	bus: EventBus,
): void {
	api.post("/internal/events", async (c) => {
		const result = InternalEventBody.safeParse(await c.req.json());
		if (!result.success) {
			throw new ValidationError("Invalid event body", result.error.issues);
		}
		bus.emit({ type: result.data.type, data: result.data.data });
		return c.body(null, 204);
	});
}

File: src/rest/internal.rs
use axum::{Router, Json, extract::State};
use axum::routing::post;
use axum::http::StatusCode;
use serde::Deserialize;
use serde_json::Value;
use crate::core::error::AppError;
use crate::rest::AppState;

pub fn routes() -> Router<AppState> {
    Router::new()
        .route("/internal/events", post(relay_event))
}

#[derive(Deserialize)]
struct RelayInput {
    #[serde(rename = "type")]
    event_type: String,
    data: Value,
}

async fn relay_event(
    State(state): State<AppState>,
    Json(input): Json<RelayInput>,
) -> Result<StatusCode, AppError> {
    let re = regex_lite::Regex::new(r"^[a-z]+\.[a-z]+$").unwrap();
    if !re.is_match(&input.event_type) {
        return Err(AppError::validation("Invalid event type", None));
    }
    state.bus.emit(input.event_type, input.data);
    Ok(StatusCode::NO_CONTENT)
}
